modules/oe_webtools_etrans/tests/src/Kernel/ETransBlockTest.php

```diff
@@ -0,0 +1,170 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\Tests\oe_webtools_etrans\Kernel;
+
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\language\Entity\ConfigurableLanguage;
+
+/**
+ * Tests the block that displays the Webtools eTrans link.
+ */
+class ETransBlockTest extends KernelTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = [
+    'system',
+    'user',
+    'language',
+    'locale',
+    'oe_webtools_etrans',
+  ];
+
+  /**
+   * The language manager.
+   *
+   * @var \Drupal\Core\Language\LanguageManagerInterface
+   */
+  protected $languageManager;
+
+  /**
+   * The service to set the default language.
+   *
+   * @var \Drupal\Core\Language\LanguageDefault
+   */
+  protected $languageDefault;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+    $this->installEntitySchema('user');
+    $this->installSchema('system', 'sequences');
+    $this->installConfig(['system']);
+    $this->languageManager = \Drupal::service('language_manager');
```

Comment:
```suggestion
    $this->languageManager =$this->container->get('language_manager');
    $this->languageDefault = $this->container->get('language.default');
```
There is big discussion about how services have to be invoked in tests https://www.drupal.org/project/drupal/issues/2066993. For Kernel test usage of $this->container is safe and this approach is used in the test below. So let's do it consistent in whole test, if you don't have objections.

Reply:
Please read again carefully the issue you linked. This is exactly why I am using `\Drupal::service()` instead of `$this->container()`.

I see now that the issue summary is not correct but this is dating from 8 years. The discussion in this issue has done a complete 180 degree flip against the original proposal, and is now advocating for the opposite. The relevant discussion starts at comment 35 (written by Berdir). This and subsequent discussion is advocating _against_ the use of `$this->container`, giving a lot of evidence and examples of problems arising from it.

So if possible I would prefer to keep it, it is the current best practice AFAIK.

Reply:
Yes, sure it can be kept as is.
 I've read whole thread. It was mentioned that $this->container shouldn't be used in functional test but can be used in kernel test. My only concern is that \Drupal::service() and $this->container() are used both in this test.

Reply:
You're right, I saw too that most comments are actually for `BrowserTestBase`, Kernel tests are safe, just as you said.

I will approve this suggestion. Thanks for catching it!